TNL-4366 Hide Timed Exams

efischer19 opened the pull request.

efischer19 commented:
https://github.com/edx/edx-proctoring/pull/288 will need to land before this PR does.

efischer19 commented:
As of the latest commit, this seems to be working correctly on my devstack. I'm going to get a sandbox set up and write some tests before asking for review.

efischer19 commented:
Sandbox is up and seems to be working. Users `audit`, `verified`, and `honor` have all completed the exam in this subsection, which was then changed to be due in the past. https://efischer19.sandbox.edx.org/courses/course-v1:EricX+test+2016/courseware/8cb2b50cc12946908da02bbd42303a3f/39f8b6dc145d4fb4b89f05d601675616/

I did run into some issues with not getting the correct `edx-proctoring` version installed as part of sandbox provisioning, but I was able to manually get it working. Something to keep an eye on for future deployments.

Next steps:
- [ ] fix currently failing tests in both PRs
- [ ] plan and implement new test coverage for this feature
- [ ] review and merge/deploy

efischer19 commented:
The latest bok choy failure is due to showing a **timed** exam where we should be showing a **proctored** exam, and it only happens on jenkins. Trying to run without my feature flags commit now to see if that fixes things.

efischer19 commented:
jenkins run all

efischer19 commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -872,61 +876,71 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
                     setEditModalValues("7/9/2014", "7/10/2014", "Lab", true);
                     selectPracticeExam("00:30");
                     // id_time_limit_div should not be hidden when practice exam is specified
-                    expect($('#id_time_limit_div')).not.toHaveClass('is-hidden"');
```
This line (and others like it) were caused me a lot of grief. Notice the extraneous trailing `"` character? Nothing ever matches that class! This test had been verifying that certain elements did not contain a fictitious class, which was always true.

In investigating what was actually going on here, I learned 2 major things:
1. The time limit, review rules, and hide-after-due options are inside of divs that have an `is-hidden` class. This is always true, even when the divs are visible! We control visibility via `show()` and `hide()`, which set the element's `display` to be `none` and `block`, respectively. I tried removing the `is-hidden` class from the elements in the template, but that broke the modal's responsiveness so I let it be for now.
2. Many of these tests needed to be calling `selectAdvancedSettings()` in order to actually open the advanced settings tab, and confirm that certain elements were visible that should have been visible.

efischer19 commented:
jenkins run lettuce

efischer19 commented:
All that remains on this PR before review/merge is to finish the new acceptance test, which is close-but-not-quite-there

efischer19 commented:
jenkins run js

efischer19 commented:
@cahrens @dianakhuang This is now ready for review. I tried to separate my changes by commit, so that may be the easiest way to view the full diff.

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
FYI @douglashall This is the platform PR corresponding to https://github.com/edx/edx-proctoring/pull/288 which you reviewed last week

cahrens commented on cms/djangoapps/contentstore/tests/test_proctoring.py:
```diff
@@ -74,6 +75,10 @@ def test_publishing_exam(self, is_time_limited, default_time_limit_minutes,
         """
 
+        hide_after_due = False
+        if not is_proctored_exam and not is_practice_exam:
```
Is it possible to just always allow "hide_after_due"? That would simplify the code, and I don't see a compelling reason to disallow it in a couple of cases.

cahrens replied:
@efischer19 and I talked-- the proctored and practice proctored exams implement their own views. Therefore the hide_after_due property is currently not respected for those exam types.

I think for proctored and practice proctored we just shouldn't pass a value of hide_after_due.

efischer19 replied:
All 3 kinds of special exams (timed, proctored, practice) will have a `hide_after_due` field, which defaults to `False`. So we're not really passing a value here so much as saying "publish an exam with the defaults".

Since this test is specifically checking `Happy path testing to see that when a course is published which contains a proctored exam, it will also put an entry into the exam tables`, I'm going to rip out the specialized logic here and just use the default value of false for all exam types.

There is/will be coverage for the functionality of `hide_after_due` including both possible values, but this test isn't the right place for it.

cahrens commented on cms/envs/common.py:
```diff
@@ -187,7 +187,7 @@
 
     # Special Exams, aka Timed and Proctored Exams
-    'ENABLE_SPECIAL_EXAMS': False,
+    'ENABLE_SPECIAL_EXAMS': True,
```
Did you intend to check this in? What are the implications?

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -670,8 +671,9 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
                             "is_practice_exam": false,
-                            "is_proctored_exam": true,
-                            "default_time_limit_minutes": 150
+                            "is_proctored_exam": false,
```
Why is is_proctored_exam being set to false?

efischer19 replied:
This test structure wasn't doing anything special related to proctored exams, so I changed it to be a regular timed exam (see https://github.com/edx/edx-proctoring/blob/master/edx_proctoring/api.py#L1828-L1843 for how this is handled by the proctoring API) in order to extend the test to cover the new feature.

cahrens replied:
OK

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -944,18 +958,21 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
                                     "is_proctored_exam": false,
-                                    "default_time_limit_minutes": 150
+                                    "default_time_limit_minutes": 150,
+                                    "hide_after_due": false,
```
Will hide_after_due always be passed back from the server, even if it is not relevant (as in this case)?

efischer19 replied:
Yes, it will. It's a field on the `ProctoredExam` model, and is stored with the exam definition.

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -872,61 +876,71 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
+                    selectAdvancedSettings();
                     setEditModalValues("7/9/2014", "7/10/2014", "Lab", true);
-                    selectTimedExam("00:30");
+                    selectTimedExam("00:30", true);
```
Do you cover the case of a timed exam with "hide_after_due" set to false?

efischer19 replied:
I do, but not in this test. Here we're only checking to see that the field is visible, so the actual value of the boolean is irrelevant. I'll change this call to not pass any value for clarity.

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -828,8 +830,9 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
+                    selectAdvancedSettings();
                     setEditModalValues("7/9/2014", "7/10/2014", "Lab", true);
-                    selectProctoredExam("02:30");
+                    selectTimedExam("02:30", true);
```
I am concerned about the loss of proctored exam coverage.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -343,31 +343,38 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
             this.$('#id_time_limit').val('00:00');
         },
-        selectSpecialExam: function (showRulesField) {
+        selectSpecialExam: function (additionalOptions) {
```
There should be documentation about what the allowed values are. The current hyphenated names seem odd to me-- if you stick with this approach, I would suggest "showRulesField" and "showHideAfterDue" (these values only control whether or not the field is shown, not the selected value of them).

Two different boolean values would be more explicit, and also a more standard pattern.

efischer19 replied:
Will switch back to boolean parameters

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -343,31 +343,38 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
         setTimedExam: function (event) {
             event.preventDefault();
-            this.selectSpecialExam(false);
+            this.selectSpecialExam(["hide-after-due"]);
```
Also cover timed exam without hide-after-due.

efischer19 replied:
This is a special case of the method commented on above - it's saying "make the `hide-after-due` field visible", which should always happen for timed exams.

cahrens replied:
Oh right-- sorry! This is not the test code.

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -69,5 +69,16 @@
                 <p class='field-message' id='review-rules-description'><%- gettext('Specify any additional rules or rule exceptions that the proctoring review team should enforce when reviewing the videos. For example, you could specify that calculators are allowed.') %></p>
             </ul>
         </div>
+        <div class='timed-exam-options-list-fields is-hidden'>
```
@chris-mike I find this UI pretty funky. Two things bother me:

1. The excessive list nesting (even though there is only a single item).
2. The fact that optionally appearing sections always appear at the end, instead of being nested within the exam type that they relate to. I would expect this checkbox to appear within the "Timed" exam section instead of at the end (since we are only supporting it for Timed exams, not Proctored or Practice Proctored exams). Similarly, I think the "Review Rules" text field should appear within the "Proctored" exam section.

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -69,5 +69,16 @@
+            <ul class="list-fields list-input timed-exam-options">
+                <li class="field-checkbox">
+                    <input type="checkbox" id="id_hide_exam_after_due_date" name="hide_after_due"  class="input input-checkbox"/>
+                    <label for="id_hide_after_due_date" class="label" aria-describedby="hide-after-due-description">
```
Put input inside of label so that you don't have to specify "for". Perhaps then you can get rid of the ID field altogether (we try to avoid IDs).

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -69,5 +69,16 @@
+                        <%- gettext('Hide Exam After Due Date') %>
+                    </label>
+                </li>
+                <p class='field-message' id='hide-after-due-description'><%- gettext('By default, submitted exams are available for review after the due date has passed. Select this option to keep exams hidden after that date.') %></p>
```
If this checkbox remains at the very end, consider adding explanatory text that this is only supported for Timed exams.

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -69,5 +69,16 @@
+                    <input type="checkbox" id="id_hide_exam_after_due_date" name="hide_after_due"  class="input input-checkbox"/>
+                    <label for="id_hide_after_due_date" class="label" aria-describedby="hide-after-due-description">
+                        <%- gettext('Hide Exam After Due Date') %>
+                    </label>
```
If there is not already a Bok Choy accessibility test (that opens this dialog and clicks on the Advanced settings tab before verifying accessibility, please add one).

cahrens commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -97,6 +97,15 @@ class ProctoringFields(object):
+    hide_after_due = Boolean(
+        display_name=_("Hide Exam Results After Due Date"),
+        help=_(
+            "This setting overrides the default behavior of showing exam results after the due date has passed."
```
Add "Currently only supported for Timed Exams".

catong replied:
@efischer19 Are we using "timed exams" to include proctored exams (which are always timed)?
If yes, should this note read "Currently supported only for timed and proctored exams" for clarity? 
Also would recommend getting rid of initial caps for "Timed Exams".

efischer19 replied:
@catong No, "Timed Exams" and "Proctored Exams" are treated separately, and the new feature only applies to the former. We refer to them with capitalized names in several places already (screenshots, also "Practice Exams" just below this line), making me think they're more proper nouns than general descriptions. Do you still want to remove the capitalization?

<img width="236" alt="screen shot 2016-04-28 at 10 20 06 am" src="https://cloud.githubusercontent.com/assets/7373924/14889065/2f1f8a16-0d2b-11e6-9f42-68f6b643bf03.png">
<img width="441" alt="screen shot 2016-04-28 at 10 19 51 am" src="https://cloud.githubusercontent.com/assets/7373924/14889064/2f1f088e-0d2b-11e6-85a3-d59d29a2d04d.png">

catong replied:
@efischer19 The instances I'm seeing of timed exam etc. with initial caps look to me like titles or labels, so initial caps are fine in those cases. But in a sentence, it would be best to lowercase "timed exams", so let's do that here. I looked on the sandbox incl. in Advanced Settings but am not sure where this text shows up?

efischer19 replied:
👍 sounds good to me, I'll get it changed. I'm not 100% sure where this is visible, it's a property field on an xmodule. I'll do some poking around and get back to you

catong replied:
Thanks, Eric. No need to spend time tracking this down, unless you really want to! :-)

cahrens commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -558,6 +558,17 @@ def change_problem_release_date_in_studio(self):
 
+    def change_problem_due_date_in_studio(self, date):
+        """
+        Sets a new start date.
```
Should be due date, not start date.

cahrens commented on common/test/acceptance/pages/studio/overview.py:
```diff
@@ -558,6 +558,17 @@ def change_problem_release_date_in_studio(self):
+        Expects date to be a string that will be accepted by the input (for example, '01/01/1970')
+        """
+        self.q(css=".subsection-header-actions .configure-button").first.click()
+        self.q(css="#due_date").fill(date)
```
Can you use the existing due_date setter?

efischer19 replied:
I tried this, it led to test failures. Going back to the approach used by `change_problem_release_date` above.

cahrens commented on common/test/acceptance/tests/lms/test_lms_courseware.py:
```diff
@@ -246,7 +249,8 @@ def test_proctored_exam_flow(self):
+    @ddt.data(True, False)
+    def test_timed_exam_flow(self, hide_after_due):
         """
         Given that I am a staff member on the exam settings section
```
Update the BDD description.

cahrens commented:
@efischer19 I'm done with my review.

cahrens commented:
Oh, but also please check with @chris-mike about whether or not the Studio outline should indicate the "hide after due" state for a timed exam.

![image](https://cloud.githubusercontent.com/assets/484484/14795317/ff8acbb2-0af5-11e6-9810-a6e0b47f73ac.png)

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -446,6 +461,7 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
             var is_proctored_exam;
             var time_limit = this.getExamTimeLimit();
             var exam_review_rules = this.$('#id_exam_review_rules').val();
+            var hide_after_due = this.$('#id_hide_exam_after_due_date').is(':checked');
```
Note that since you always send this checked value (regardless of whether or not it is a timed exam), you may end up sending "true" for a proctored exam if the author selected "hide after due" while Timed exam was selected. This is probably fine since it is just ignored by proctored exams, but think through if there are any issues with this.

efischer19 replied:
Yup, this is intentional. This makes it so the value is persisted, so if you change the exam type away from timed, then back, the value you had previously will remain.

chris-mike commented:
I think it's a good idea to Indicate the "hide after due" status in the studio course outline.  In terms of UI, I recommend adding it as a new line under the exam check box.

Some other considerations:

1. After talking with @shelbystack, it seems that hiding the timed exam ALSO removes exam information from a student's progress (their completed and graded exam appears as 0% in the progress page).  If this is a bug, then I would  not recommend a change. But if this 0% is a feature change, I'd recommend changing language around what is said.  Something like:  "Exam and results - hidden after due date."

2. Is there a reason why the check mark icons differ?  Do they change once the exam is complete?

cahrens commented:
@chris-mike -- it is not true that hiding the exam will impact the progress page. The student will see their exam grade there, they just won't be able to view the exam itself. @efischer19 can follow up with Shelby to clarify how this works (since it does not follow our usual has_access pattern).

efischer19 commented:
I've followed up on the TNL ticket, the progress page is unaffected by this change so we'll keep the wording the same.

The icons are different, should we fix that? The released status check is either [fa-check-square-o](http://fortawesome.github.io/Font-Awesome/icon/check-square-o/) or [fa-clock-o](http://fortawesome.github.io/Font-Awesome/icon/clock-o/) depending on whether the subsection is released or not. The grading label is then always [fa-check](http://fortawesome.github.io/Font-Awesome/icon/check/). These are all defined here: https://github.com/edx/edx-platform/blob/master/cms/templates/js/course-outline.underscore#L143-L190

It'd be pretty easy to standardize on a single checkmark style, if that would make things flow better.

efischer19 commented:
@chris-mike Here's what I've added to indicate the hide after due status on the course outline

<img width="968" alt="screen shot 2016-04-26 at 2 17 15 pm" src="https://cloud.githubusercontent.com/assets/7373924/14829488/aefbcb46-0bb9-11e6-9558-ce7083921d3c.png">

chris-mike commented:
@efischer19  just a quick note about language. You want to say that "exam will be hidden after due date," not that it will "remain" hidden. Right?

chris-mike commented:
@efischer19 one MORE thing...  color contrast fails wcag aa. Can you make it #707070.

efischer19 commented:
Re: the language - the current behavior is for a (submitted) timed exam to be hidden until the due date passes, then become visible after. So with this setting exams will remain hidden, versus becoming visible.

I was just working on fixing the contrast on the settings modal, I'll get this display as well.

efischer19 commented:
@chris-mike I ended up going with the variable [gray-d1](https://github.com/edx/edx-platform/blob/master/cms/static/sass/_variables.scss#L64) (#66666), is that okay, or should I force #707070 explicitly? This has a ratio of 5.7, and passes AA level http://leaverou.github.io/contrast-ratio/#%23666-on-%23ffffff

chris-mike commented:
Hey, @efischer19 : #66666 is fine on #f2f2f2 (bkgd). In future, you can check here: http://webaim.org/resources/contrastchecker/. We need to be WCAG AA compliant, at minimum

efischer19 commented:
jenkins run js

efischer19 commented:
@cahrens @dianakhuang Ready for re-review! I've addressed all the prior review comments, added an a11y test (which is failing on jenkins for some reason, but passes locally), and fixed up the `timed-examination-preference-editor` template to be more logical and accessible.

My sandbox is up to date, I'm going to keep investigating why jenkins is mad at me.

cahrens commented on cms/djangoapps/contentstore/tests/test_proctoring.py:
```diff
@@ -85,7 +89,8 @@ def test_publishing_exam(self, is_time_limited, default_time_limit_minutes,
             due=datetime.now(UTC) + timedelta(minutes=default_time_limit_minutes + 1),
-            exam_review_rules="allow_use_of_paper"
+            exam_review_rules="allow_use_of_paper",
+            hide_after_due=False,
```
In this test case, add a parameter for hide_after_due and make it true in the timed exam case (so we have some coverage for it being true).

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -872,50 +876,66 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
-                    expect($('#id_time_limit_div')).not.toHaveClass('is-hidden"');
+
+                    // time limit should be visible, review rules and hide after due should be hidden
+                    expect($('.field-time-limit').is(':visible')).toBe(true);
```
Nit: since these 3 things are often checked together, create a helper method for verifying their state.

cahrens commented on cms/static/js/spec/views/pages/course_outline_spec.js:
```diff
@@ -971,18 +996,21 @@ define(["jquery", "common/js/spec_helpers/ajax_helpers", "common/js/components/u
                                     "is_proctored_exam": false,
-                                    "default_time_limit_minutes": 10
+                                    "default_time_limit_minutes": 10,
+                                    "hide_after_due": true,
```
Add coverage for when hide_after_due is False for a timed exam.

cahrens commented on cms/static/js/views/modals/course_outline_modals.js:
```diff
@@ -332,41 +332,47 @@ define(['jquery', 'backbone', 'underscore', 'gettext', 'js/views/baseview',
         setPracticeExam: function (event) {
             event.preventDefault();
-            this.selectSpecialExam(false);
+            this.selectSpecialExam();
```
Nit: I would err on the side of being more explicit (passing false, false).

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -177,6 +177,17 @@ if (is_proctored_exam) {
+                    <p>
+                        <% if (xblockInfo.get('hide_after_due')) { %>
+                            <i class="icon fa fa-eye-slash"></i>
+                            <span class="status-hide-after-due-value"> <%= gettext("Exam will remain hidden after due date") %> </span>
```
Should be <%- to be XSS safe (in case of nefarious translation, and our best practice is to always escape unless it is known to be HTML).

Same below.

cahrens commented on cms/templates/js/course-outline.underscore:
```diff
@@ -177,6 +177,17 @@ if (is_proctored_exam) {
+                            <span class="status-hide-after-due-value"> <%= gettext("Exam will remain hidden after due date") %> </span>
+                        <% } else { %>
+                            <i class="icon fa fa-eye"></i>
+                            <span class="status-hide-after-due-value"> <%= gettext("Exam will be visible after due date") %> </span>
```
@chris-mike Do you like always having a message visible? I could also see only including this visibility piece of information in the case of hide_after_due being True (the non-default case).

FYI @catong these changes will need to be documented.

cahrens commented on cms/templates/js/timed-examination-preference-editor.underscore:
```diff
@@ -1,73 +1,65 @@
+                </label>
+            </li>
+            <li class="field-radio">
+                <label for="id_timed_exam" class="label">
```
Can remove this "for".

cahrens commented:
@efischer19 This is looking good-- thanks for all the updates! Just a few more things to tidy up.

dianakhuang commented:
👍 from me.

efischer19 commented on common/test/acceptance/tests/studio/test_studio_settings.py:
```diff
@@ -508,3 +508,65 @@ def test_studio_settings_page_a11y(self):
+            browser = 'firefox'
+
+        with patch.dict(os.environ, {'SELENIUM_BROWSER': browser}):
+            super(StudioSubsectionSettingsA11yTest, self).setUp(is_staff=True)
```
@jzoldak or @clytwynec, is this sort of browser switching acceptable to merge into master? Without this, the test will always fail due to phantomjs 1.9.8 not handling `bind()` correctly.

- [Failing test](https://build.testeng.edx.org/job/edx-platform-accessibility-pr/17087/)
- [Browser screenshot, can't load Course Outline](https://build.testeng.edx.org/job/edx-platform-accessibility-pr/17087/artifact/edx-platform/test_root/log/common.test.acceptance.tests.studio.test_studio_settings.StudioSubsectionSettingsA11yTest.test_special_exams_menu_a11y.png)
- [Browser log of js error](https://build.testeng.edx.org/job/edx-platform-accessibility-pr/17087/artifact/edx-platform/test_root/log/common.test.acceptance.tests.studio.test_studio_settings.StudioSubsectionSettingsA11yTest.test_special_exams_menu_a11y_browser.log/*view*/)
- [Offending js in master](https://github.com/edx/edx-platform/blob/master/cms/static/js/views/course_outline.js#L182-L192)

jzoldak replied:
@benpatterson Tagging for your opinion here.

benpatterson replied:
Makes sense to me. We're doing a workaround and have logged a ticket pointing out that we shouldn't have to. I also wonder if we should pivot to installing phantomjs via npm, but that itself would be another story.

efischer19 commented:
@cahrens I've addressed your latest rounds of comments. Can you give this another look?

Remaining before merge:
- [ ] @cahrens thumbs
- [ ] testeng signoff on the "swap browsers for this test" code for the new a11y test
- [ ] talk with @catong to finalize wording, get documentation written
- [ ] remove feature flag commit
- [ ] squash commits
- [ ] successful test run of everything on jenkins

efischer19 commented:
jenkins run lettuce

efischer19 commented:
jenkins run bokchoy

efischer19 commented:
jenkins run lettuce

cahrens commented:
👍
